lang/c++/impl/ZstdCodecWrapper.cc

```diff
@@ -0,0 +1,112 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef ZSTD_CODEC_AVAILABLE
+
+#include "ZstdCodecWrapper.hh"
+#include "Exception.hh"
+
+#include <zstd.h>
+
+namespace avro {
+
+std::vector<char> ZstdCodecWrapper::compress(const std::vector<char> &uncompressed) {
+    if (cctx_ == nullptr) {
+        initCCtx();
+    }
+    // Pre-allocate buffer for compressed data
+    size_t max_compressed_size = ZSTD_compressBound(uncompressed.size());
+    std::vector<char> compressed(max_compressed_size);
+
+    // Compress the data using ZSTD block API
+    size_t compressed_size = ZSTD_compress(
+        compressed.data(), max_compressed_size,
+        uncompressed.data(), uncompressed.size(),
+        ZSTD_CLEVEL_DEFAULT);
+
+    if (ZSTD_isError(compressed_size)) {
+        throw Exception("ZSTD compression error: {}", ZSTD_getErrorName(compressed_size));
+    }
+    compressed.resize(compressed_size);
+    return compressed;
+}
+
+std::string ZstdCodecWrapper::decompress(const std::vector<char> &compressed) {
+    if (dctx_ == nullptr) {
+        initDCtx();
+    }
+    std::string uncompressed;
+    // Get the decompressed size
+    size_t decompressed_size = ZSTD_getFrameContentSize(
+        reinterpret_cast<const char *>(compressed.data()), compressed.size());
+    if (decompressed_size == ZSTD_CONTENTSIZE_ERROR) {
+        throw Exception("ZSTD: Not a valid compressed frame");
+    } else if (decompressed_size == ZSTD_CONTENTSIZE_UNKNOWN) {
+        // Stream decompress the data
+        ZSTD_inBuffer in{compressed.data(), compressed.size(), 0};
+        std::vector<char> tmp(ZSTD_DStreamOutSize());
+        ZSTD_outBuffer out{tmp.data(), tmp.size(), 0};
+        size_t ret;
+        do {
+            out.pos = 0;
+            ret = ZSTD_decompressStream(dctx_, &out, &in);
+            if (ZSTD_isError(ret)) {
+                throw Exception("ZSTD decompression error: {}", ZSTD_getErrorName(ret));
+            }
+            uncompressed.append(tmp.data(), out.pos);
+        } while (ret != 0);
+    } else {
+        // Batch decompress the data
+        uncompressed.resize(decompressed_size);
+        size_t result = ZSTD_decompress(
+            uncompressed.data(), decompressed_size,
+            reinterpret_cast<const char *>(compressed.data()), compressed.size());
+
+        if (ZSTD_isError(result)) {
+            throw Exception("ZSTD decompression error: {}", ZSTD_getErrorName(result));
+        }
+        if (result != decompressed_size) {
+            throw Exception("ZSTD: Decompressed size mismatch: expected {}, got {}",
+                            decompressed_size, result);
+        }
+    }
+    return uncompressed;
+}
+
+void ZstdCodecWrapper::initCCtx() {
```

Comment:
This is not idiomatic C++. Initialization should happen in the constructor. This protects from being accidentally called more than once. There is nothing shared between the wrapper used for compression and the one for decompression. It will be far cleaner if we have two different wrappers - one for compression and the other for decompression. Then we don't have to worry about `cctx_`/`dctx_` being `null`.